lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java

```diff
@@ -0,0 +1,235 @@
+ */
+
+package org.apache.lucene.util.hnsw;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.apache.lucene.index.KnnGraphValues;
+import org.apache.lucene.index.VectorValues;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.apache.lucene.util.VectorUtil.dotProduct;
+import static org.apache.lucene.util.VectorUtil.squareDistance;
+
+/**
+ * Navigable Small-world graph. Provides efficient approximate nearest neighbor
+ * search for high dimensional vectors.  See <a href="https://doi.org/10.1016/j.is.2013.10.006">Approximate nearest
+ * neighbor algorithm based on navigable small world graphs [2014]</a> and <a
+ * href="https://arxiv.org/abs/1603.09320">this paper [2018]</a> for details.
+ *
+ * This implementation is actually more like the one in the same authors' earlier 2014 paper in that
+ * there is no hierarchy (just one layer), and no fanout restriction on the graph: nodes are allowed to accumulate
+ * an unbounded number of outbound links, but it does incorporate some of the innovations of the later paper, like
+ * using a priority queue to perform a beam search while traversing the graph. The nomenclature is a bit different
+ * here from what's used in those papers:
+ *
+ * <h3>Hyperparameters</h3>
+ * <ul>
+ *   <li><code>numSeed</code> is the equivalent of <code>m</code> in the 2012 paper; it controls the number of random entry points to sample.</li>
+ *   <li><code>beamWidth</code> in {@link HnswGraphBuilder} has the same meaning as <code>efConst</code> in the 2016 paper. It is the number of
+ *   nearest neighbor candidates to track while searching the graph for each newly inserted node.</li>
+ *   <li><code>maxConn</code> has the same meaning as <code>M</code> in the later paper; it controls how many of the <code>efConst</code> neighbors are
+ *   connected to the new node</li>
+ *   <li><code>fanout</code> the fanout parameter of {@link VectorValues#search(float[], int, int)}
+ *   is used to control the values of <code>numSeed</code> and <code>topK</code> that are passed to this API.
+ *   Thus <code>fanout</code> is like a combination of <code>ef</code> (search beam width) from the 2016 paper and <code>m</code> from the 2014 paper.
+ *   </li>
+ * </ul>
+ *
+ * <p>Note: The graph may be searched by multiple threads concurrently, but updates are not thread-safe. Also note: there is no notion of
+ * deletions. Document searching built on top of this must do its own deletion-filtering.</p>
+ */
+public final class HnswGraph {
+
+  // each entry lists the neighbors of a node, in node order
+  private final List<List<Neighbor>> graph;
+
+  HnswGraph() {
+    graph = new ArrayList<>();
+    graph.add(new ArrayList<>());
+  }
+
+  /**
+   * Searches for the nearest neighbors of a query vector.
```

Comment:
This is expected to normally be used just during indexing?  Maybe explain that in the javadoc?  I.e. at search time, users should use the other `search` method that pulls neighbors from the index and recurses.

Reply:
No this method is intended for use both when indexing and when searching. I guess this method could go in some other class (KnnGraphValues??) since it doesn't rely on Hnsw in any direct way (it only needs an abstract source of vectors and graph values). But it does use Neighbor/Neighbors and is heavily used *by* HnswGraphBuilder, so it seems natural to have it here.